### ISSUE TITLE
Overloading operators ++ and --

### ISSUE THREAD

**Issue by jakobkummerow** [littledan/proposal-operator-overloading#8]
I'd like to point out that there's a crucial difference for `++` and `--` between what the native operators do and what one can customize with anything that desugars to functions. Consider:

```
var x = MyThing();
foo.bar = x;
baz.qux = x;
foo.bar++;
console.log(baz.qux);
```

Presented like this example, it seems fairly obvious that `foo.bar++` modifies `baz.qux` too, contrary to what would happen if `x` were a plain number. If I modify the example, it gets more confusing:

```
// Assume Vector has overloaded operators.
const kZero = new Vector([0, 0, 0]);

class Particle {
  constructor(...) {
    this.speed = kZero;
  }
  accelerate() {
    this.speed++;
  }
}
```

Because of this footgun, and because of the mental model that (all other) operators are essentially syntactic sugar for static functions that don't modify their inputs, I would argue that overloading `++` and `--` is not useful.

(Full disclosure: personally I dislike operator overloading in general, but I concede that that's a personal opinion, whereas I think the above is a real concern.)

**Comment by littledan** [comment]
I agree that `++` and `--` shouldn't modify the object, but rather return a new one, and then the system would reset the binding appropriately. This should probably be documented better. For the same reason, I omitted overloading `+=`, although Python includes it.

Does this resolve your concern, or am I misunderstanding?

**Comment by Zachiah** [comment]
Why couldn't all of the "sugar" operators just translate to there equivelents so in this example
```javascript
// Assume Vector has overloaded operators.
const kZero = new Vector([0, 0, 0]);

class Particle {
  constructor(...) {
    this.speed = kZero;
  }
  accelerate() {
    this.speed++;
  }
}
```

`this.speed++` just means `this.speed = this.speed + 1`
it could also work the same for += as in `this.speed += 4` is equivelent to `this.speed = this.speed + 4`. If it worked that way you wouldn't need to define `+=` or `++` but just `+`.

**Comment by jhmaster2000** [comment]
In general this would be having these operators be implicitly overloaded by their more primitive counterparts, I proposed a full table on how I believe this should work over at #53 which includes the `++` and `--` suggested here.
